# task-03/chapter-12(webscraping)


---

## Libraries Used
- `webbrowser`
- `sys`
- `pyperclip`

### Functionality  
- **`webbrowser`**: Used to open the default web browser with the Google Maps URL.  
- **`sys`**: Handles command-line arguments.  
- **`pyperclip`**: Accesses the system clipboard for copied text.

---

In [8]:
import webbrowser, sys, pyperclip

if len(sys.argv) < 1:
    address = ' '.join(sys.argv[1:])

else:
    address = pyperclip.paste()

webbrowser.open("https://www.google.com/maps/place/" + address)

True

---

## Libraries Used
- `requests`

### Functionality
1. **`requests.get(url)`**  
   - Fetches the content of the URL.  
   - URL used: `https://automatetheboringstuff.com/files/rj.txt`.  

2. **`res.status_code`**  
   - Checks the HTTP status code of the response.  
   - `requests.codes.ok` confirms the request was successful (status code 200).  

3. **`res.text`**  
   - Contains the full content of the fetched file as a string.  
   - `len(res.text)` gives the total number of characters.  

4. **`res.text[:250]`**  
   - Displays the first 250 characters of the fetched content.  

--- 

In [9]:
import requests
res = requests.get('https://automatetheboringstuff.com/files/rj.txt')
print(type(res)) 

print(res.status_code == requests.codes.ok)  

print(len(res.text))  
print(res.text[:250])


<class 'requests.models.Response'>
True
178978
The Project Gutenberg EBook of Romeo and Juliet, by William Shakespeare

This eBook is for the use of anyone anywhere at no cost and with
almost no restrictions whatsoever.  You may copy it, give it away or
re-use it under the terms of the Projec


## Checking for errors:
---
## Libraries Used
- requests
### Functionality
1. requests.get(url)

- Fetches the content of the specified URL.
- URL used: https://inventwithpython.com/page_that_does_not_exist.

2. res.raise_for_status()

-  Raises an HTTPError if the HTTP request returned an unsuccessful status code.
- Useful for error handling when dealing with URLs that may not exist.
---

In [11]:
res = requests.get('https://inventwithpython.com/page_that_does_not_exist')
res.raise_for_status()

HTTPError: 404 Client Error: Not Found for url: https://inventwithpython.com/page_that_does_not_exist

## Saving Downloaded Files to the Hard Drive

---

## Libraries Used
- `requests`

### Functionality
1. **Fetch Content from URL**  
   - Uses `requests.get()` to fetch the content of the text file from the URL:  
     `https://automatetheboringstuff.com/files/rj.txt`.

2. **Error Handling**  
   - Uses `res.raise_for_status()` to check if the HTTP request was successful.  
   - Raises an exception if an error occurred (e.g., 404 Not Found).

3. **Save Content to File**  
   - Opens a file named `RomeoAndJuliet.txt` in binary write mode (`'wb'`).  
   - Writes the content in chunks of 100,000 bytes to handle large files efficiently.  
   - Closes the file after writing is complete.

**File Created**:  
- `RomeoAndJuliet.txt`  

**Key Method**:  
- `res.iter_content(100000)`: Reads the response content in chunks of 100,000 bytes.  

---



In [14]:
import requests

res = requests.get('https://automatetheboringstuff.com/files/rj.txt')
res.raise_for_status()
playFile = open('RomeoAndJuliet.txt', 'wb')
for chunk in res.iter_content(100000):
        playFile.write(chunk)

playFile.close()